### ISSUE TITLE
SFTP connectivity issues

### ISSUE THREAD

**Issue by Ugluth** [proftpd/proftpd#805]
Hello, I have recently setup a ProFTPD server so that our customers can upload their files. The files are uploaded automatically by different software. The control on the client software is quite limited. So after colaborating with them, I was able to get them to try to connect to a test server where I had trace enabled, in hopes that it would solve the problem. However I can't see anything wrong from the trace log. So I thought of posting here in case I can get some support.

I used to run ProFTPD version 1.3.5a and the specific customer could upload to the server (more or less similar config). On the new server that runs version 1.3.5e they can't connect anymore. I'm pretty sure that this is a configuration issue, but I really can't figure out what. The client is trying to connect via SFTP. So gonna post the relevant configuration.

[sftp-trace.log](https://pastebin.com/MUJU2XGk)

**proftpd.conf**
```
TraceLog /var/log/proftpd/trace.log
Trace DEFAULT:10

# Includes DSO modules
Include /etc/proftpd/modules.conf

UseIPv6                         off
IdentLookups                    off
TraceLog /var/log/proftpd/sftp-trace.log
Trace scp:20 sftp:20 ssh2:20
<Global>
        ServerIdent on "Welcome!"
        User                            proftpd
        Group                           www-data
        AuthOrder                       mod_sql.c*
        DefaultRoot                     ~
        RequireValidShell               off
        Umask                           0006    0007
        AllowOverwrite                  on
        <Limit SITE_CHMOD>
                DenyAll
        </Limit>
        PassivePorts                  49152 65534
</Global>

ServerName                      "my.domain.com"
ServerType                      standalone
DeferWelcome                    off
MultilineRFC2228                on
DefaultServer                   on
ShowSymlinks                    off
TimeoutNoTransfer               600
TimeoutStalled                  600
TimeoutIdle                     1200
DisplayLogin                    welcome.msg
DisplayChdir                    .message true
ListOptions                     "-l"
DenyFilter                      \*.*/

Port                            21
MasqueradeAddress               X.X.X.X
<IfModule mod_quotatab.c>
	QuotaEngine off
</IfModule>
<IfModule mod_ratio.c>
	Ratios off
</IfModule>
<IfModule mod_delay.c>
	DelayEngine on
</IfModule>
<IfModule mod_ctrls.c>
	ControlsEngine        off
	ControlsMaxClients    5
	ControlsLog           /var/log/proftpd/controls.log
	ControlsInterval      5
	ControlsSocket        /var/run/proftpd/proftpd.sock
</IfModule>
<IfModule mod_ctrls_admin.c>
	AdminControlsEngine off
</IfModule>
Include /etc/proftpd/sql.conf
Include /etc/proftpd/tls.conf
Include /etc/proftpd/conf.d/
```

**sftp.conf**
```
<IfModule mod_sftp.c>
        <VirtualHost 0.0.0.0>
                Port 2222
                ServerName                      "my.domain.com"
                <Limit LOGIN>
                        AllowAll
                </Limit>
                SFTPEngine on
                SFTPLog /var/log/proftpd/sftp.log
                SFTPHostKey /etc/ssh/ssh_host_rsa_key
                SFTPHostKey /etc/ssh/ssh_host_dsa_key
                SFTPAuthMethods password
                SFTPCompression delayed
                SFTPClientMatch .* channelWindowSize 64MB
        </VirtualHost>
</IfModule>
```

**proftpd -V**
```
$ proftpd -V
Compile-time Settings:
  Version: 1.3.5e (maint)
  Platform: LINUX [Linux 4.15.0-1037-aws x86_64]
  Built: Mon Feb 5 2018 23:27:47 UTC
  Built With:
    configure  '--build=x86_64-linux-gnu' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--libexecdir=${prefix}/lib/x86_64-linux-gnu' '--disable-maintainer-mode' '--disable-dependency-tracking' '--prefix=/usr' '--with-includes=/usr/include/postgresql:/usr/include/mysql' '--mandir=/usr/share/man' '--sysconfdir=/etc/proftpd' '--localstatedir=/run' '--libexecdir=/usr/lib/proftpd' '--enable-sendfile' '--enable-facl' '--enable-dso' '--enable-autoshadow' '--enable-ctrls' '--with-modules=mod_readme' '--enable-ipv6' '--enable-nls' '--enable-memcache' '--with-lastlog=/var/log/lastlog' '--enable-pcre' '--disable-strip' '--build' 'x86_64-linux-gnu' '--with-shared=mod_unique_id:mod_site_misc:mod_load:mod_ban:mod_quotatab:mod_sql:mod_sql_mysql:mod_sql_postgres:mod_sql_sqlite:mod_sql_odbc:mod_dynmasq:mod_quotatab_sql:mod_ldap:mod_quotatab_ldap:mod_ratio:mod_tls:mod_rewrite:mod_radius:mod_wrap:mod_wrap2:mod_wrap2_file:mod_wrap2_sql:mod_quotatab_file:mod_quotatab_radius:mod_facl:mod_ctrls_admin:mod_copy:mod_deflate:mod_ifversion:mod_tls_memcache:mod_geoip:mod_exec:mod_sftp:mod_sftp_pam:mod_sftp_sql:mod_shaper:mod_sql_passwd:mod_ifsession' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-IzaTCF/proftpd-dfsg-1.3.5e=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-IzaTCF/proftpd-dfsg-1.3.5e=. -fstack-protector-strong -Wformat -Werror=format-security'

  CFLAGS: -g -O2 -fdebug-prefix-map=/build/proftpd-dfsg-IzaTCF/proftpd-dfsg-1.3.5e=. -fstack-protector-strong -Wformat -Werror=format-security -Wall
  LDFLAGS: -L$(top_srcdir)/lib -Wl,-Bsymbolic-functions -Wl,-z,relro  -L/usr/lib/x86_64-linux-gnu -L/usr/lib/x86_64-linux-gnu
  LIBS:  -lpcreposix -lpcre -lssl -lcrypto -lcap  -lmemcached -lmemcachedutil  -lpam -lsupp -lcrypt -ldl

  Files:
    Configuration File:
      /etc/proftpd/proftpd.conf
    Pid File:
      /run/proftpd.pid
    Scoreboard File:
      /run/proftpd.scoreboard
    Header Directory:
      /usr/include/proftpd
    Shared Module Directory:
      /usr/lib/proftpd
  Features:
    + Autoshadow support
    + Controls support
    + curses support
    - Developer support
    + DSO support
    + IPv6 support
    + Largefile support
    + Lastlog support
    + Memcache support
    + ncurses support
    + NLS support
    + OpenSSL support
    + PCRE support
    + POSIX ACL support
    + Shadow file support
    + Sendfile support
    + Trace support

  Tunable Options:
    PR_TUNABLE_BUFFER_SIZE = 1024
    PR_TUNABLE_DEFAULT_RCVBUFSZ = 8192
    PR_TUNABLE_DEFAULT_SNDBUFSZ = 8192
    PR_TUNABLE_GLOBBING_MAX_MATCHES = 100000
    PR_TUNABLE_GLOBBING_MAX_RECURSION = 8
    PR_TUNABLE_HASH_TABLE_SIZE = 40
    PR_TUNABLE_NEW_POOL_SIZE = 512
    PR_TUNABLE_SCOREBOARD_BUFFER_SIZE = 80
    PR_TUNABLE_SCOREBOARD_SCRUB_TIMER = 30
    PR_TUNABLE_SELECT_TIMEOUT = 30
    PR_TUNABLE_TIMEOUTIDENT = 10
    PR_TUNABLE_TIMEOUTIDLE = 600
    PR_TUNABLE_TIMEOUTLINGER = 30
    PR_TUNABLE_TIMEOUTLOGIN = 300
    PR_TUNABLE_TIMEOUTNOXFER = 300
    PR_TUNABLE_TIMEOUTSTALLED = 3600
    PR_TUNABLE_XFER_SCOREBOARD_UPDATES = 10
```

Any help or guidance regarding this issue would be really appreciated! If you need any further info please ask. Thanks in advance!

**Comment by Ugluth** [comment]
I forgot to mention some info. The old version of ProFTPD (1.3.5a) was running on Ubuntu 16.04 and the new server where I'm facing this issue (ProFTPD version 1.3.5e) is running Ubuntu 18.04.

What I also did is bring up the old FTP server where the same customer could upload files fine. I enabled trace logging on that one as well and compared the 2 trace logs.

For the new version of the server it seems to be failing after this trace message:
```
2019-05-06 19:09:03,265 [3117] <ssh2:12>: computing DH key
2019-05-06 19:09:03,265 [3117] <signal:9>: handling SIGSEGV (signal 11)
```
While on the old FTP server it continues normally: 
```
2019-05-07 13:24:03,732 [22656] <ssh2:12>: computing DH key
2019-05-07 13:24:03,733 [22656] <ssh2:19>: waiting for max of 600 secs while polling socket 1 using select(2)
2019-05-07 13:24:03,733 [22656] <ssh2:3>: sent SSH_MSG_KEX_DH_REPLY (31) packet (640 bytes)
2019-05-07 13:24:03,733 [22656] <ssh2:9>: sending NEWKEYS message to client
```

This seems a bit similar to issue #789 
Openssl version on the new server (1.3.5e): `OpenSSL 1.1.0g  2 Nov 2017`
Openssl version on the old server (1.3.5a): `OpenSSL 1.0.2g  1 Mar 2016`

Again if you need any further info plz ask.

**Comment by ErwanGa** [comment]
Hello
Same as #789

The current releases (1.3.5e and 1.3.6) have an issue around  DSA_SIG_get0 usage that cause segmentation fault when proftpd is compiled on top of openssl 1.1.x

The fix is already on master branch and 1.3.6 maintenance branch 

@Castaglia:  It could be great to get an official release/tag. As is such patches will be deploy all fixes in distributions


Cheers
Erwan

**Comment by Castaglia** [comment]
With ProFTPD 1.3.6b and 1.3.7rc2 being released, with this fix, I'm considering this as addressed.  Thanks!
